File: samples/client/3_1_0_unit_test/java/src/main/java/org/openapijsonschematools/client/schemas/validation/MinPropertiesValidator.java
package org.openapijsonschematools.client.schemas.validation;

import org.openapijsonschematools.client.exceptions.ValidationException;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.Map;

public class MinPropertiesValidator implements KeywordValidator {
    public final int minProperties;

    public MinPropertiesValidator(int minProperties) {
        this.minProperties = minProperties;
    }

    @Override
    public @Nullable PathToSchemasMap validate(
        ValidationData data
    ) {
        if (!(data.arg() instanceof Map<?, ?> mapArg)) {
            return null;
        }
        if (mapArg.size() < minProperties) {
            throw new ValidationException("Value " + mapArg + " is invalid because has < the minProperties of " + minProperties);
        }
        return null;
    }
}

File: samples/client/petstore/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/MinPropertiesValidator.kt
package org.openapijsonschematools.client.schemas.validation

import org.openapijsonschematools.client.exceptions.ValidationException

class MinPropertiesValidator : KeywordValidator {
    @Throws(ValidationException::class)
    override fun validate(
        data: ValidationData
    ): PathToSchemasMap? {
        val minProperties: Int = data.schema.minProperties ?: return null
        if (data.arg !is Map<*, *>) {
            return null
        }
        if (data.arg.size < minProperties) {
            throw ValidationException("Value ${data.arg} is invalid because has < the minProperties of $minProperties")
        }
        return null
    }
}
